### PR TITLE
Some refactoring around intrinsic type checking

### PR DESCRIPTION
**Pull request by oli-obk**
So... This PR went a bit overboard. I wanted to make the `rustc_peek` intrinsic safe (cc @ecstatic-morse ), and remembered a long-standing itch of mine. So I made that huge `&str` match for the intrinsic name a match on `Symbol`s (so basically `u32`s). This is unlikely to have a positive perf effect, even if it likely has better codegen (intrinsics are used rarely, mostly once in their wrapper), so it's mostly a consistency thing since other places actually match on the symbol name of the intrinsics.

### CODE REVIEW

**Comment by oli-obk** [comment]
Ok... so something really fishy is going on... the symbol id of the symbol we are matching on is 54, which, when manually counting the elements in https://github.com/rust-lang/rust/blob/master/src/librustc_span/symbol.rs#L90 is the correct index. But here comes the thing I don't get. `sym::Try.to_u32()` gives me 799, which makes no sense whatsoever.

**Comment by oli-obk** [comment]
Nevermind... I was using `sym::Try` instead of `kw::Try`, the latter is `try` while the former is `Try` for the trait name.

File: src/librustc_typeck/check/intrinsic.rs [hunk 1]
```diff
@@ -67,15 +67,17 @@ fn equate_intrinsic_type<'tcx>(
 }
 
 /// Returns `true` if the given intrinsic is unsafe to call or not.
-pub fn intrinsic_operation_unsafety(intrinsic: &str) -> hir::Unsafety {
+pub fn intrinsic_operation_unsafety(intrinsic: Symbol) -> hir::Unsafety {
+    use rustc_span::symbol::sym::*;
```
Comment (ecstatic-morse) [inline src/librustc_typeck/check/intrinsic.rs:71]: A glob import of `sym` pollutes the local namespace with a lot of lower-case constants. For this reason, it's not done anywhere else in the compiler. I realize that it's kind of a pain here, but we should uphold this example. It's more relevant in the other big match statement, which also has other code.

File: src/librustc_typeck/check/intrinsic.rs [hunk 2]
```diff
@@ -303,7 +302,7 @@ pub fn check_intrinsic_type(tcx: TyCtxt<'_>, it: &hir::ForeignItem<'_>) {
                 )
             }
 
-            "try" => {
+            rustc_span::symbol::kw::Try => {
```
Comment (ecstatic-morse) [inline src/librustc_typeck/check/intrinsic.rs:305]: Good catch. Having `sym::Try` and `kw::Try` mean different things is a bit subtle. I don't see how we could do better though.

**Review by ecstatic-morse — COMMENTED** [review]
r=me with glob imports removed and replaced with `sym::xxx`.

File: src/librustc_typeck/check/intrinsic.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_typeck/check/intrinsic.rs]: yea... I knew that to be the right thing... but... laziness XD

**Comment by Manishearth** [comment]
Is there a chance this PR is massively slowing down CI? It was a part of two rollups (https://github.com/rust-lang/rust/pull/73880, https://github.com/rust-lang/rust/pull/73864) that timed out, and while this is not the only suspect, it's worth investigating.

**Comment by oli-obk** [comment]
I didn't notice anything locally... but let's see

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
It would be good to check each new symbol (e.g. `sym::assume`) for any other occurrences of the corresponding string in the code base (e.g. "assume"). There might be some string operations that can now be changed to symbol operations.

**Comment by oli-obk** [comment]
> It would be good to check each new symbol (e.g. `sym::assume`) for any other occurrences of the corresponding string in the code base (e.g. "assume"). There might be some string operations that can now be changed to symbol operations.

definitely, but I wanted to get this PR merged first, as any further work on moving to symbols will be much simpler and smaller PRs.

**Comment by oli-obk** [comment]
perf is clean, it's not likely to be this PR

@bors r=ecstatic-morse rollup

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/73939#issuecomment-652654320

**Comment by oli-obk** [comment]
@bors r+

It's not this PR, some other PR didn't rebase and bless. I'll figure out the problematic PR

**Comment by oli-obk** [comment]
I'm guessing it was https://github.com/rust-lang/rust/pull/73622/files#diff-c4db7f9b504f4fbce6ba0b53e7810379 but they updated their ui files since then

**Comment by oli-obk** [comment]
I'll try to figure it out later, but some PR included a miri submodule update, even though it should not, idk which one yet.

**Comment by oli-obk** [comment]
ah it was https://github.com/rust-lang/rust/pull/72983 which already knows about it

**Pull request merged**